net/sched/p4tc/p4tc_cmds.c

```diff
@@ -923,24 +952,53 @@ static void free_op_ACT(struct p4tc_cmd_operate *ope,
 	if (p) {
 		struct tcf_idrinfo *idrinfo = p->idrinfo;
 
-		if (refcount_dec_and_mutex_lock(&p->tcfa_refcnt,
-						&idrinfo->lock)) {
-			idr_remove(&idrinfo->action_idr, p->tcfa_index);
-			mutex_unlock(&idrinfo->lock);
+		if (dec_act_refs) {
+			atomic_dec(&p->tcfa_bindcnt);
 
-			if (p->ops->cleanup)
-				p->ops->cleanup(p);
+			if (refcount_dec_and_mutex_lock(&p->tcfa_refcnt,
+							&idrinfo->lock)) {
+				idr_remove(&idrinfo->action_idr, p->tcfa_index);
+				mutex_unlock(&idrinfo->lock);
 
-			gen_kill_estimator(&p->tcfa_rate_est);
-			_free_tcf(p);
-		}
+				if (p->ops->cleanup)
+					p->ops->cleanup(p);
 
-		atomic_dec(&p->tcfa_bindcnt);
+				gen_kill_estimator(&p->tcfa_rate_est);
+				_free_tcf(p);
+			}
+		}
 	}
 
 	return _free_operation(ope, extack);
 }
 
+static inline int opnd_is_assignable(struct p4tc_cmd_operand *kopnd,
+				     struct netlink_ext_ack *extack)
+{
+	switch (kopnd->oper_type) {
+	case P4TC_OPER_CONST:
+		NL_SET_ERR_MSG_MOD(extack, "Unable to assign to constant");
+		return -EINVAL;
+	case P4TC_OPER_RES:
+		NL_SET_ERR_MSG_MOD(extack, "Unable to assign to results field");
+		return -EINVAL;
+	case P4TC_OPER_PARAM:
+		NL_SET_ERR_MSG_MOD(extack, "Unable to assign to parameter");
+		return -EINVAL;
+	case P4TC_OPER_META:
+		if (kopnd->pipeid == 0 && kopnd->immedv == P4TC_KERNEL_META_SKBIIF) {
+			NL_SET_ERR_MSG_MOD(extack,
+					   "Assignment to kernel.metadata.iif is forbidden");
+			return -EINVAL;
+		}
+		break;
```

Comment:
I still think this could be a simple flag check like so:
if (kopnd->constraints & CUD) {
   NL_SET_ERR_VARAGS_MSG_MOD(extack, "Not allowed to SET on %s", kopnd->type);
   return -EINVAL;
}

or something along those lines..

Reply:
Done here: https://github.com/p4tc-dev/linux-p4tc-pub/pull/45/files#diff-f70ae2b5731743e0d086cae79f9de3fd014258802d40560d8ca53b4ed1261ec3R982